### PR TITLE
refactor(inference): correct Message types to reflect multimodal request content

### PR DESCRIPTION
**Pull request by HONGJICAI**
## Summary

This is a **type-correctness refactor**, not a feature enablement. Multimodal (vision) input already flows through the broker end-to-end today, because the inference path reads the request as raw bytes (`io.ReadAll` in `proxy/proxy.go`), hashes those bytes directly for signing, and forwards them unchanged to the provider. The `RequestBody`/`Message` structs in `chatbot.go` are not unmarshaled anywhere in the request path — they are effectively documentation.

The problem: the existing `Message.Content string` is an **incorrect type** for request messages. OpenAI-compatible clients may send `content` as either a plain string or an array of content parts (text + image_url). Any future code that does `json.Unmarshal` into the old `RequestBody` would silently drop multimodal content or fail outright. This PR fixes the type definitions so that future inspection/billing/logging code built on top of them won't break multimodal requests.

## Changes

- Introduce `MessageContent` with custom `MarshalJSON`/`UnmarshalJSON` that accepts both a plain string and an OpenAI multimodal array (text + image_url parts).
- Split `Message` into:
  - `RequestMessage` — `Content` is `MessageContent` (string or multimodal array).
  - `ResponseMessage` — `Content` remains `string`, since LLM responses are always plain text.
- Update `RequestBody.Messages` and `Choice.Message` to use the correct variant.
- Add unit tests covering string, multimodal, null, and round-trip (un)marshaling.

## Scope / non-goals

- **No runtime behavior change.** The proxy path still forwards raw bytes; signing still hashes raw bytes. Text-only and multimodal requests behave exactly as before this PR.
- No changes to proxy, signing, billing, or response handling.

## How to verify

- `go test ./api/inference/internal/ctrl/...` — unit tests for the new (un)marshaling.
- Text-only chat requests continue to work unchanged (no path reads these structs).
- Multimodal vision requests continue to work unchanged (already did pre-PR).

## Test plan

- [ ] Unit tests pass (`go test ./api/inference/internal/ctrl/...`)
- [ ] Text-only chat requests still work end-to-end
- [ ] Multimodal requests with base64 `image_url` are forwarded correctly (same as before)
- [ ] Response parsing (always plain string) is unaffected

Refs #421

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: api/inference/internal/ctrl/chatbot.go [hunk 1]
```diff
@@ -56,8 +56,71 @@ type ChatSignature struct {
...
+}
+
+// ImageURL holds an image reference in an OpenAI-compatible content part.
+type ImageURL struct {
+	URL    string `json:"url"`
+	Detail string `json:"detail,omitempty"`
+}
+
+func (mc MessageContent) MarshalJSON() ([]byte, error) {
+	if mc.Parts != nil {
```
Comment (HONGJICAI) [inline api/inference/internal/ctrl/chatbot.go]: [MEDIUM] **`MarshalJSON`: nil check vs length check**

Using `mc.Parts != nil` means an empty but non-nil slice (`[]ContentPart{}`) would marshal as `[]` (empty JSON array) rather than `""`. This can happen if, for example, a client sends `{"content": []}` — `UnmarshalJSON` produces a non-nil empty slice, and re-marshaling it yields `[]` instead of the arguably more correct `""`.

Consider using `len(mc.Parts) > 0` which is also more idiomatic Go:

```go
func (mc MessageContent) MarshalJSON() ([]byte, error) {
	if len(mc.Parts) > 0 {
		return json.Marshal(mc.Parts)
	}
	return json.Marshal(mc.Text)
}
```

This also makes the zero-value `MessageContent{}` consistently marshal as `\"\"` regardless of how it was constructed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: api/inference/internal/ctrl/chatbot.go [hunk 2]
```diff
@@ -56,8 +56,71 @@ type ChatSignature struct {
...
+			return err
+		}
+		// Also populate Text with concatenated text parts for convenience.
+		var sb strings.Builder
+		for _, p := range mc.Parts {
+			if p.Type == "text" {
+				sb.WriteString(p.Text)
+			}
+		}
+		mc.Text = sb.String()
```
Comment (HONGJICAI) [inline api/inference/internal/ctrl/chatbot.go:108]: [LOW] **Multiple text parts concatenated without separator**

When a multimodal content array contains multiple `text` parts, they're concatenated directly without any delimiter. For example, parts `[\"Hello\", \"World\"]` become `\"HelloWorld\"` instead of `\"Hello World\"` or `\"Hello\nWorld\"`.

In practice the OpenAI API typically has a single text part per message, so this is unlikely to cause real issues. But if you want to be defensive, consider adding a space or newline separator:

```go
for i, p := range mc.Parts {
	if p.Type == \"text\" {
		if i > 0 && sb.Len() > 0 {
			sb.WriteString(\" \")
		}
		sb.WriteString(p.Text)
	}
}
```

Minor — only matters if `Text` is used for downstream token estimation or display.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: api/inference/internal/ctrl/chatbot.go [hunk 3]
```diff
@@ -56,8 +56,71 @@ type ChatSignature struct {
...
+		return nil
+	}
+	// null or other — treat as empty.
+	return nil
```
Comment (HONGJICAI) [inline api/inference/internal/ctrl/chatbot.go:112]: [LOW] **Silent acceptance of unexpected JSON types**

The fallback case silently returns `nil` for any JSON token that isn't a string (`\"`), array (`[`), or null. This means a JSON number like `123` or an object like `{\"key\":\"val\"}` would produce an empty `MessageContent` with no error, which could mask upstream bugs.

Consider returning an error for truly unexpected types:

```go
// null — treat as empty.
if bytes.Equal(data, []byte(\"null\")) {
	return nil
}
return fmt.Errorf(\"unsupported content type: %s\", string(data[:min(len(data), 20)]))
```

This is defensive — the OpenAI API only sends strings, arrays, or null for `content`, so it may never trigger in practice.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: api/inference/internal/ctrl/chatbot.go [hunk 4]
```diff
@@ -56,8 +56,71 @@ type ChatSignature struct {
...
+	// Parts holds the content when it is a multimodal array.
+	Parts []ContentPart
+}
+
+// ContentPart represents one element of a multimodal content array.
+type ContentPart struct {
+	Type     string    `json:"type"`
+	Text     string    `json:"text,omitempty"`
+	ImageURL *ImageURL `json:"image_url,omitempty"`
+}
```
Comment (HONGJICAI) [inline api/inference/internal/ctrl/chatbot.go:76]: [LOW] **Future extensibility: other content part types**

The OpenAI API also supports `input_audio` content parts (and may add more in the future). Since `ContentPart` only has explicit fields for `text` and `image_url`, any other part type would lose its payload data during unmarshal → remarshal.

This is fine for now since:
1. The PR scope is vision support specifically.
2. The broker forwards the raw `[]byte` request body to providers, so the `RequestBody` struct is used for inspection, not reconstruction.

Just noting for awareness — if `RequestBody` is ever used to re-serialize and forward requests, unsupported part types would be silently dropped.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: api/inference/internal/ctrl/chatbot_test.go [hunk 5]
```diff
@@ -432,3 +433,183 @@ func TestTeeServiceSigningRoundTrip(t *testing.T) {
...
+	if err := json.Unmarshal([]byte(input), &chunk); err != nil {
+		t.Fatalf("unmarshal error: %v", err)
+	}
+	if len(chunk.Choices) != 1 {
+		t.Fatalf("Choices length = %d, want 1", len(chunk.Choices))
+	}
+	if chunk.Choices[0].Message.Content != "The image shows a cat." {
+		t.Errorf("Message.Content = %q", chunk.Choices[0].Message.Content)
+	}
+}
```
Comment (HONGJICAI) [inline api/inference/internal/ctrl/chatbot_test.go:615]: :+1: Excellent test coverage — string, multimodal, null, marshal, round-trip, and response message deserialization are all well covered. The `TestRequestBody_MultimodalRoundTrip` test is especially good at validating the full end-to-end flow with mixed string and multimodal messages.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by HONGJICAI — COMMENTED** [review]
## Review: feat(inference): support multimodal (vision) input in Message types

**Overall: Good PR.** Clean design, solid tests, safe rename.

### Summary

This PR adds multimodal (vision) input support by introducing a `MessageContent` type with custom JSON marshal/unmarshal that handles both plain string and OpenAI array content formats. Splitting `Message` into `RequestMessage` and `ResponseMessage` is the right approach — it keeps request flexibility while preserving type safety on the response path.

### Verification

- **Build**: Compiles cleanly (`go build ./...` — no errors)
- **Tests**: All pass (`go test ./inference/internal/ctrl/...` — 0.666s)
- **Rename safety**: Confirmed no external references to the old `ctrl.Message` type exist outside this package. The `RequestBody.Messages` field is never iterated in Go code — requests are forwarded as raw `[]byte`, so the type change is purely for inspection/parsing purposes.

### Findings

| Severity | Issue |
|----------|-------|
| **MEDIUM** | `MarshalJSON` uses `mc.Parts != nil` — an empty non-nil slice marshals as `[]` instead of `""`. Prefer `len(mc.Parts) > 0`. |
| **LOW** | Multiple text parts concatenated without separator (no space/newline between them). |
| **LOW** | `UnmarshalJSON` silently accepts unexpected JSON types (numbers, objects) as empty content. |
| **LOW** | `ContentPart` only supports `text` and `image_url` — other OpenAI part types (e.g., `input_audio`) would lose data on round-trip. Fine for now since raw bytes are forwarded, but worth noting for future. |

### Positives

- Excellent test coverage (6 new tests covering all serialization paths)
- Clean type separation between request and response messages
- Good use of `strings.Builder` for text concatenation
- Well-documented types with godoc comments and examples

None of the issues are blocking. The MEDIUM item is the most actionable — easy one-line fix for robustness.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Ravenyjh — APPROVED** [review]
